File: Python/Two Pointers/15. 3Sum.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        index = 0
        while index < len(nums) and nums[index] <= 0:
            if index == 0 or (nums[index-1] != nums[index]):
                self.twoSum(nums, index, res)
            index += 1
        return res
        
    def twoSum(self, nums, index, res):
        left, right = index+1, len(nums)-1
        while left < right:
            s = nums[index] + nums[left] + nums[right]
            if s == 0:
                res.append([nums[index], nums[left], nums[right]])
                left += 1
                right -= 1
                while left < right and nums[left]==nums[left-1]: left += 1
            elif s < 0:
                left += 1
            else:
                right -= 1

File: C++/Two Pointers/K Sum Series/15. 3Sum.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size() && nums[i] <= 0; i++) {
            if (i == 0 || nums[i-1] != nums[i]) {
                twoSum(nums, i, res);
            }
        }
        return res;
    }
    
    void twoSum(vector<int>& nums, int index, vector<vector<int>>& res) {
        int left = index+1, right = nums.size()-1;
        while (left < right) {
            int sum = nums[index] + nums[left] + nums[right];
            if (sum == 0) {
                res.emplace_back(vector<int>{nums[index], nums[left], nums[right]});
                left++; right--;
                while (left < right && nums[left] == nums[left-1]) left++;
            }
            else if (sum < 0) {
                left++;
            }
            else {
                right--;
            }
        }
    }
};